clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
-        let ret_adjust = if match_type(cx, ret_ty, &paths::RESULT) {
-            returns
-        } else {
-            returns / 2
-        };
-
-        if cc + divergence < match_arms + short_circuits {
-            report_cc_bug(
-                cx,
-                cc,
-                match_arms,
-                divergence,
-                short_circuits,
-                ret_adjust,
-                span,
-                body.id().hir_id,
-            );
-        } else {
-            let mut rust_cc = cc + divergence - match_arms - short_circuits;
-            // prevent degenerate cases where unreachable code contains `return` statements
-            if rust_cc >= ret_adjust {
-                rust_cc -= ret_adjust;
-            }
-            if rust_cc > self.limit.limit() {
-                span_help_and_lint(
-                    cx,
-                    COGNITIVE_COMPLEXITY,
-                    span,
-                    &format!("the function has a cognitive complexity of {}", rust_cc),
-                    "you could split it up into multiple smaller functions",
-                );
-            }
-        }
-    }
-}
+                helper.visit_block(&fn_block);
+
+                let fn_score = helper.score as u64;
+                let score_limit = self.limit.limit();
 
-impl<'a, 'tcx> LateLintPass<'a, 'tcx> for CognitiveComplexity {
-    fn check_fn(
-        &mut self,
-        cx: &LateContext<'a, 'tcx>,
-        _: intravisit::FnKind<'tcx>,
-        _: &'tcx FnDecl,
-        body: &'tcx Body,
-        span: Span,
-        hir_id: HirId,
-    ) {
-        let def_id = cx.tcx.hir().local_def_id_from_hir_id(hir_id);
-        if !cx.tcx.has_attr(def_id, "test") {
-            self.check(cx, body, span);
+                if fn_score > score_limit {
+                    span_help_and_lint(
+                        cx,
+                        COGNITIVE_COMPLEXITY,
+                        item.span,
+                        &format!("the function has a cognitive complexity of {}", fn_score),
+                        "you could split it up into multiple smaller functions",
+                    );
+                }
+            }
         }
     }
 
-    fn enter_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn enter_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.push_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
-    fn exit_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn exit_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.pop_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
 }
 
-struct CCHelper<'a, 'tcx: 'a> {
-    match_arms: u64,
-    divergence: u64,
-    returns: u64,
-    short_circuits: u64, // && and ||
-    cx: &'a LateContext<'a, 'tcx>,
+struct CoCHelper {
+    current_nesting: u64,
+    score: f64,
```

Comment:
I think this should be a `u64`, it's only ever incremented by `u64` values or `1`

Reply:
I have an idea. If we add multiplicative modifiers, we could just append a 0 to all base scores and then we can do things like `score += (12 * delta) / 10` to add `delta x 1.2`.

I think it's very probable that we do add multiplicative modifiers. This path would however let us clean up the code while keeping that a possibility :)

Reply:
That has been implemented now :)